x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -352,76 +358,78 @@ public String toString() {
-                                        fileName,
-                                        TimeValue.timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis()
-                                    )
-                                );
-                            }
-
-                            @Override
-                            public void onFailure(Exception e) {
-                                logger.trace(
-                                    () -> new ParameterizedMessage(
-                                        "failed to warm cache for [{}] part [{}/{}]",
-                                        fileName,
-                                        part,
-                                        numberOfParts
-                                    ),
-                                    e
-                                );
-                            }
-
-                            @Override
-                            public void onAfter() {
-                                if (countDown.countDown()) {
-                                    IOUtils.closeWhileHandlingException(input);
-                                }
-                            }
-                        });
-                    }
-                } catch (IOException e) {
-                    logger.trace(() -> new ParameterizedMessage("failed to warm cache for [{}]", fileName), e);
-                }
+            logger.debug("{} warming shard cache for [{}] files", shardId, queue.size());
+
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
+                prewarmNextFile(executor, queue);
             }
         }
     }
 
+    private void prewarmNextFile(final Executor executor, final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> queue) {
+        executor.execute(new AbstractRunnable() {
+            @Override
+            protected void doRun() throws Exception {
+                final BlobStoreIndexShardSnapshot.FileInfo file = queue.poll(0L, TimeUnit.MILLISECONDS);
+                if (file == null) {
+                    return;
+                }
+
+                final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
+                assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
+
+                final int numberOfParts = Math.toIntExact(file.numberOfParts());
+                final GroupedActionListener<Void> listener = new GroupedActionListener<>(
+                    ActionListener.runAfter(
+                        ActionListener.wrap(() -> IOUtils.closeWhileHandlingException(input)),
+                        () -> prewarmNextFile(executor, queue)
```

Comment:
IIUC this means we wait for all parts of this file to be warmed before starting on the next file, which sounds like it might under-utilise the available threads. Could we use something cleverer than a `Queue<FileInfo>` to track the progress of the work in terms of parts that keeps all the threads busy until the very end?

Reply:
I agree this is not optimal. The alternative I see is to enqueue file parts instead of file and then have workers polling each part to warm. I pushed 1aea98042c48781b63edd56be0e4b17ceeca07f8, it's a bit less readable IMO but still more than what I did in previous iterations. Let me know what you think.

Reply:
Yes I like that solution, thanks for the extra iteration. I don't think it's made a material difference to readability but I do think it's clearer now that the tasks are the warming of the individual parts rather than the files; also I just noticed that previously we would bound the worker pool size by the number of files rather than the number of parts, which may have resulted in less concurrency than we wanted -- this is now addressed.